src/main/java/com/emc/pravega/perf/PravegaTransactionWriterWorker.java

```diff
@@ -29,35 +29,39 @@
 public class PravegaTransactionWriterWorker extends PravegaWriterWorker {
     private final AtomicReference<Transaction<String>> transaction;
     private final int transactionsPerCommit;
-    private final AtomicInteger eventCount;
+    private final AtomicInteger transEventCount;
 
-    PravegaTransactionWriterWorker(int sensorId, int eventsPerSec,
+    PravegaTransactionWriterWorker(int sensorId, int eventsPerWorker,
                                    int secondsToRun, boolean isRandomKey,
                                    int messageSize, Instant start,
-                                   ClientFactory factory, PerfStats stats,
-                                   String streamName,
-                                   long totalEvents,
-                                   int transactionsPerCommit) {
+                                   PerfStats stats, String streamName,
+                                   ClientFactory factory, int transactionsPerCommit) {
 
-        super(sensorId, eventsPerSec, secondsToRun, isRandomKey,
-                messageSize, start, factory, stats, streamName, totalEvents);
+        super(sensorId, eventsPerWorker, secondsToRun, isRandomKey,
+                messageSize, start, stats, streamName, factory);
 
         this.transactionsPerCommit = transactionsPerCommit;
-        eventCount = new AtomicInteger(0);
+        transEventCount = new AtomicInteger(0);
         transaction = new AtomicReference<Transaction<String>>(producer.beginTxn());
     }
 
     @Override
-    public CompletableFuture writeData(String key, String data) throws TxnFailedException, IllegalStateException {
+    public CompletableFuture writeData(String key, String data) {
         final Transaction<String> curTrans = transaction.get();
 
-        curTrans.writeEvent(key, data);
-        if (eventCount.incrementAndGet() >= transactionsPerCommit) {
-            eventCount.set(0);
-            curTrans.commit();
-            if (!transaction.compareAndSet(curTrans, producer.beginTxn())) {
-                throw new IllegalStateException("WriteData called on the same PravegaTransactionWriterWorker from two threads in parallel.");
+        try {
+            curTrans.writeEvent(key, data);
+            synchronized (this) {
```

Comment:
This solves the race condition. Now you can simplify transEventCount can be a regular integer guardedBy("this"). And the if block that throws the illegal state can be removed.

Additionally you can also do the same with transaction. It can be made into a normal member guarded by 'this', and you don't need the atomic reference any more.

Reply:
yes , its done